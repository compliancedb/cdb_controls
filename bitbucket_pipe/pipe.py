import sys
from bitbucket_pipes_toolkit import Pipe, get_logger

import cdb.control_junit
import cdb.create_release
import cdb.put_artifact
import cdb.put_artifact_image
import cdb.put_pipeline
from cdb.bitbucket import get_bitbucket_repo_url, adapt_put_artifact_image_env_variables, \
    adapt_create_release_variables, adapt_control_junit_env_variables, adapt_put_artifact_env_variables

sys.path.append('/app/')
import cdb.cdb_utils
import cdb.bitbucket
from cdb.cdb_utils import set_artifact_sha_env_variable_from_file_or_image


schema = {
    'CDB_COMMAND': {'type': 'string', 'required': True},
    'CDB_PIPELINE_DEFINITION': {'type': 'string', 'required': True},
    'CDB_API_TOKEN': {'type': 'string', 'required': True},
    # PUT Pipeline Variables
    #   ...none...
    # Put Artifact Variables
    'CDB_IS_COMPLIANT': {'type': 'boolean', 'required': False},
    'CDB_ARTIFACT_GIT_URL': {'type': 'string', 'required': False},
    'CDB_ARTIFACT_GIT_COMMIT': {'type': 'string', 'required': False},
    'CDB_CI_BUILD_URL': {'type': 'string', 'required': False},
    'CDB_BUILD_NUMBER': {'type': 'string', 'required': False},
    'CDB_ARTIFACT_SHA': {'type': 'string', 'required': False},
    'CDB_ARTIFACT_FILENAME': {'type': 'string', 'required': False},

}

logger = get_logger()


class BitbucketPipe(Pipe):
    def run(self):
        super().run()

        logger.info('Executing the pipe...')

        command = self.get_variable('CDB_COMMAND')
        pipeline_definition_file = self.get_variable('CDB_PIPELINE_DEFINITION')
        api_token = self.get_variable('CDB_API_TOKEN')

        print("Command: " + command)
        print("Pipeline: " + pipeline_definition_file)
        print("API_TOKEN: " + api_token)

        if command == "put_pipeline":
            cdb.put_pipeline.put_pipeline(pipeline_definition_file)
        if command == "put_artifact":
            adapt_put_artifact_env_variables()
            cdb.put_artifact.put_artifact(pipeline_definition_file)
        if command == "put_artifact_image":
            adapt_put_artifact_image_env_variables()
            cdb.put_artifact_image.put_artifact_image(pipeline_definition_file)
        if command == "control_junit":
            adapt_control_junit_env_variables()
            cdb.control_junit.control_junit(pipeline_definition_file)
        if command == "control_bitbucket_pr":
            set_artifact_sha_env_variable_from_file_or_image()
            cdb.bitbucket.put_bitbucket_pull_request(pipeline_definition_file)
        if command == "create_release":
            adapt_create_release_variables()
            cdb.create_release.create_release(pipeline_definition_file)

        self.success(message="Success!")


if __name__ == '__main__':
    pipe = BitbucketPipe(pipe_metadata='/pipe.yml', schema=schema)
    pipe.run()
