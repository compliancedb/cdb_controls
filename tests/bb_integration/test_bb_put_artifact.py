from pipe import BitbucketPipe, schema

from tests.utils import AutoEnvVars, CDB_DRY_RUN, verify_approval


def test_required_env_vars(capsys, mocker):
    env = {
        "CDB_COMMAND": "put_artifact",
        "CDB_PIPELINE_DEFINITION": "tests/data/pipefile.json",
        "CDB_API_TOKEN": "SOME_RANDOM_TOKEN",
        "CDB_ARTIFACT_FILENAME": "src/data/robocop.xml",
        "BITBUCKET_COMMIT": "abc50c8a53f79974d615df335669b59fb56a4ed3",
        "BITBUCKET_BUILD_NUMBER": "127",
        "BITBUCKET_WORKSPACE": "acme",
        "BITBUCKET_REPO_SLUG": "road-runner"
    }
    sha = "ddcdaef69c676c2466571d3288880d559ccc2032b258fc5e73f99a103db462ee"
    set_env_vars = {
        'CDB_ARTIFACT_GIT_URL': 'https://bitbucket.org/acme/road-runner/commits/abc50c8a53f79974d615df335669b59fb56a4ed3',
        'CDB_ARTIFACT_GIT_COMMIT': 'abc50c8a53f79974d615df335669b59fb56a4ed3', 'CDB_BUILD_NUMBER': '127',
        'CDB_CI_BUILD_URL': 'https://bitbucket.org/acme/road-runner/addon/pipelines/home#!/results/127',
        'CDB_ARTIFACT_SHA': 'ddcdaef69c676c2466571d3288880d559ccc2032b258fc5e73f99a103db462ee'
    }
    with AutoEnvVars({**CDB_DRY_RUN, **env}, set_env_vars):
        pipe = BitbucketPipe(pipe_metadata='/pipe.yml', schema=schema)
        mocker.patch('cdb.cdb_utils.calculate_sha_digest_for_file', return_value=sha)
        pipe.run()
    verify_approval(capsys)


