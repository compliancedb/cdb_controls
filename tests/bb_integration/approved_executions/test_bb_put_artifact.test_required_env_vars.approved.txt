Command: put_artifact
Pipeline: tests/data/pipefile.json
API_TOKEN: SOME_RANDOM_TOKEN
Getting SHA for artifact: src/data/robocop.xml
Calculated digest: ddcdaef69c676c2466571d3288880d559ccc2032b258fc5e73f99a103db462ee
Publish artifact to ComplianceDB
CDB_IS_COMPLIANT: False
Putting this payload:
{
    "build_url": "https://bitbucket.org/acme/road-runner/addon/pipelines/home#!/results/127",
    "commit_url": "https://bitbucket.org/acme/road-runner/commits/abc50c8a53f79974d615df335669b59fb56a4ed3",
    "description": "Created by build 127",
    "filename": "src/data/robocop.xml",
    "git_commit": "abc50c8a53f79974d615df335669b59fb56a4ed3",
    "is_compliant": false,
    "sha256": "ddcdaef69c676c2466571d3288880d559ccc2032b258fc5e73f99a103db462ee"
}
To this url: https://app.compliancedb.com/api/v1/projects/merkely/test-pipefile/artifacts/
DRY RUN: Put not sent
[32mâœ” Success![0m
